jose-b64/src/serde/bytes.rs

```diff
@@ -10,57 +10,57 @@ use core::marker::PhantomData;
 use core::ops::{Deref, DerefMut};
 use core::str::FromStr;
 
-use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
-use zeroize::{Zeroize, Zeroizing};
+use base64ct::{Base64UrlUnpadded, Encoding};
+use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
 
-use super::{Config, UrlSafe};
+use crate::stream::Error;
 
 /// A serde wrapper for base64-encoded bytes.
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
-pub struct Bytes<T = Box<[u8]>, C = UrlSafe> {
+pub struct Bytes<T = Box<[u8]>, E = Base64UrlUnpadded> {
     buf: T,
-    cfg: PhantomData<C>,
+    cfg: PhantomData<E>,
 }
 
-impl<T: Zeroize, C> Zeroize for Bytes<T, C> {
+impl<T: crate::Zeroize, E> crate::Zeroize for Bytes<T, E> {
     fn zeroize(&mut self) {
         self.buf.zeroize()
     }
 }
 
-impl<T: Debug, C> Debug for Bytes<T, C> {
+impl<T: Debug, E> Debug for Bytes<T, E> {
     fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
         f.debug_tuple("Bytes").field(&self.buf).finish()
     }
 }
 
-impl<T, C> Deref for Bytes<T, C> {
+impl<T, E> Deref for Bytes<T, E> {
```

Comment:
I'm kind of torn with the `Deref` and `DerefMut` implementation, look at the [std docs](https://doc.rust-lang.org/nightly/std/ops/trait.Deref.html):

> [...] Deref should only be implemented for smart pointers to avoid confusion.

Reply:
I think the only question is how broadly we interpret `smart pointer`. For example, `Zeroizing` uses `Deref` while not being a smart pointer in the strictest sense. In the looser sense, `Bytes` is a smart pointer in the same way `Zeroizing` is (i.e. a wrapper around a contiguous collection of bytes).

Reply:
Yeah, I'm not sure if it is alright or if it isn't. We might have to ask some other rust folks. For example, we have a similar thing in our implementation: <https://github.com/minkan-chat/jose/blob/jws-refactor/src/base64_url.rs#L75>

Reply:
FWIW, I'd consider both this application and `Zeroizing` to be "smart pointer" cases.

To me a smart pointer is a special container that is generic around some contained data it owns and mediates access to. Having a special `Drop` impl is definitely one of the intended use cases.